functionalization: make view_copy outputs always contiguous

bdhirsh opened the pull request:
This fixes an issue with mobile: The output of view_copy ops should always be contiguous.

Later, we can consider adding optional arguments to the `view_copy()` functions to let you explicitly say what the contiguity of the output can be (e.g. channels_last)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85747

bdhirsh committed:
functionalization: make view_copy outputs always contiguous

[ghstack-poisoned]

ezyang reviewed (approved):
How do you know that these spots are enough? Can you come up with a more comprehensive testing strategy that asserts that the internal tensors of functionalization are always contiguous?

larryliu0820 commented:
I guess we can rely on tags to filter out "view_copy" ops and run test on all of them. That could be a separate PR?

ngimel commented:
I don't think we want this behavior by default on non-mobile. E.g. channels-last inputs better remain channels-last, same goes for transposes.

bdhirsh commented:
@ngimel what do you think of adding an optional memory format kwarg to the existing view copy ops
```
at::Tensor permute_copy(at::Tensor self, IntArrayRef dims, c10::MemoryFormat mem_format=c10::MemoryFormat::Preserve);
```

Where functionalization will (by default) blat out `view_copy` ops with `mem_format=c10::MemoryFormat::contiguous`

That shouldn't regress behavior for anyone using view_copy operators directly today. using the functionalization API, `functionalize(f, remove="mutations_and_view")` would be a behavior change, but mobile is the only major user of that API for now. And if you have a channels_last model, where you want all intermediates/outputs to be in channels_last format, you could specify `functionalize(f, remove="mutations_and_views", memory_format=torch.channels_last)` (exact API bikeshedding probably not necessary immediately).

ngimel commented:
Yeah that sounds reasonable

bdhirsh commented:
> E.g. channels-last inputs better remain channels-last, same goes for transposes.

Hmm @ngimel - I'm not actually sure what should happen when e.g. you have a 2d tensor - "channels last contiguous" doesn't really make sense in that case, right?

Would it be reasonable to say that the behavior we want is something along the lines of:
```
if (input.is_contiguous(c10::MemoryFormat::channels_last)) return output.clone(c10::MemoryFormat::channels_last);
if (input.is_contiguous(c10::MemoryFormat::channels_last_3d)) return output.clone(c10::MemoryFormat::channels_last_3d);
return output.clone(c10::MemoryFormat::contiguous);
```

Does that sound like a better solution to you than letting the user pass in a memory format?

bdhirsh commented:
I see - I'll try that out. Thanks!

bdhirsh committed:
Update on "functionalization: make view_copy outputs always contiguous"

This fixes an issue with mobile: The output of view_copy ops should always be contiguous.

Later, we can consider adding optional arguments to the `view_copy()` functions to let you explicitly say what the contiguity of the output can be (e.g. channels_last)




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: make view_copy outputs always contiguous"

This fixes an issue with mobile: The output of view_copy ops should always be contiguous.

Later, we can consider adding optional arguments to the `view_copy()` functions to let you explicitly say what the contiguity of the output can be (e.g. channels_last)




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: make view_copy outputs always contiguous"

This fixes an issue with mobile: The output of view_copy ops should always be contiguous.

Later, we can consider adding optional arguments to the `view_copy()` functions to let you explicitly say what the contiguity of the output can be (e.g. channels_last)




[ghstack-poisoned]

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -659,23 +659,31 @@ def forward(self, a_1):
-    _reshape_alias_copy_3 = torch.ops.aten._reshape_alias_copy.default(view_copy_3, [2, 4], [4, 1]);  view_copy_3 = None
-    select_copy_1 = torch.ops.aten.select_copy.int(_reshape_alias_copy_3, 0, 0);  _reshape_alias_copy_3 = None
-    add_2 = torch.ops.aten.add.Tensor(select_copy_1, _unsafe_view);  select_copy_1 = _unsafe_view = None
+    _reshape_alias_copy_3 = torch.ops.aten._reshape_alias_copy.default(transpose_copy_2, [8], [1]);  transpose_copy_2 = None
```
These looks correct - before we would (incorrectly) decompose reshape() down the path where we needed to clone, and now we don't (because the input is guaranteed contiguous).

bdhirsh commented:
@pytorchbot merge -g
